### PR TITLE
News App

### PR DESCRIPTION
**Pull request by SagarAjudiya**
**News App**

- Category wise news
- Bookmark news
- Check network connection

### CODE REVIEW

File: app/src/main/java/com/simform/news/data/model/News.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    @SerializedName("articles")
+    val articles: List<Article>,
+    @SerializedName("status")
+    val status: String,
+    @SerializedName("totalResults")
+    val totalResults: Int
+) : Parcelable
+
+@Parcelize
+data class Article(
```
Comment (krupal-simformsolutions) [inline app/src/main/java/com/simform/news/data/model/News.kt:18]: Seperate the class

File: app/src/main/java/com/simform/news/data/model/News.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    val publishedAt: String? = null,
+    @SerializedName("title")
+    val title: String? = null,
+    @SerializedName("url")
+    val url: String? = null,
+    @SerializedName("urlToImage")
+    val urlToImage: String? = null
+) : Parcelable {
+    companion object {
+        val bookmarkList = ArrayList<Article>()
```
Comment (krupal-simformsolutions) [inline app/src/main/java/com/simform/news/data/model/News.kt:35]: why companion object

File: app/src/main/java/com/simform/news/repository/SourcesRepository.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import com.simform.news.network.ApiClient
+import com.simform.news.util.URL
+
+class SourcesRepository {
+
+    suspend fun getSources(): SourceResponse? {
+        val response = ApiClient.apiInterface.getSources("sources", URL.apiKey)
+        val result = response.body()
+        return result
+    }
```
Comment (krupal-simformsolutions) [inline app/src/main/java/com/simform/news/repository/SourcesRepository.kt:13]: handle failure cases

File: app/src/main/java/com/simform/news/ui/bookmark/view/BookMarkFragment.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        super.onViewCreated(view, savedInstanceState)
+        initViews()
+    }
+
+    private fun initViews() {
+        if (Article.bookmarkList.isEmpty()) {
+            binding.animBookmark.visibility = View.VISIBLE
+        } else {
+            binding.animBookmark.visibility = View.INVISIBLE
+        }
```
Comment (krupal-simformsolutions) [inline app/src/main/java/com/simform/news/ui/bookmark/view/BookMarkFragment.kt:48]: avoid static variables

File: app/src/main/java/com/simform/news/ui/headlines/view/BottomDialogFragment.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                dismiss()
+            }
+
+            tvBrowser.setOnClickListener {
+                startActivity(
+                    Intent(
+                        Intent.ACTION_VIEW,
+                        Uri.parse(binding.article?.url)
+                    )
+                )
```
Comment (krupal-simformsolutions) [inline app/src/main/java/com/simform/news/ui/headlines/view/BottomDialogFragment.kt:73]: what if url is empty

File: app/src/main/java/com/simform/news/ui/headlines/view/FilterAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        fun defaultBg() {
+            binding.cvFilter.background = ContextCompat.getDrawable(binding.cvFilter.context, R.drawable.bg_capsule_unselected)
+            binding.cvFilter.strokeWidth = 0
+        }
+
+        fun selectedBg() {
+            binding.cvFilter.background = ContextCompat.getDrawable(binding.cvFilter.context, R.drawable.bg_capsule_selected)
+            binding.cvFilter.strokeWidth = 4
```
Comment (krupal-simformsolutions) [inline app/src/main/java/com/simform/news/ui/headlines/view/FilterAdapter.kt:32]: we can handle it by selector drawable

File: app/src/main/java/com/simform/news/ui/newsdetails/view/ArticleFragment.kt [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+        parentFragmentManager.setFragmentResultListener(
+            "article",
+            viewLifecycleOwner
+        ) { _, bundle ->
+            val res = bundle.getParcelable<Article>("article")
+            Log.d("res", "initViews: $res")
+            binding.apply {
+                article = res
+            }
```
Comment (krupal-simformsolutions) [inline app/src/main/java/com/simform/news/ui/newsdetails/view/ArticleFragment.kt:84]: unnessosory scoping function

File: app/src/main/java/com/simform/news/ui/newsdetails/view/ArticleFragment.kt [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+//        dataViewModel.dataList.observe(viewLifecycleOwner) { dataList ->
+//            binding.apply {
+//                article = dataViewModel
+//            }
+//            Log.d("Res", "initViews: $dataList")
```
Comment (krupal-simformsolutions) [inline app/src/main/java/com/simform/news/ui/newsdetails/view/ArticleFragment.kt:91]: remove commented code

File: app/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -2,6 +2,10 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools">
 
+    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
+    <uses-permission android:name="android.permission.INTERNET"/>
+    <uses-permission android:name="android.permission.CHANGE_NETWORK_STATE"/>
+
     <application
         android:allowBackup="true"
```
Comment (priyanshudesai) [inline app/src/main/AndroidManifest.xml:10]: why used AllowBackup?

File: app/src/main/AndroidManifest.xml [hunk 10]
```diff
@@ -12,16 +16,19 @@
...
             android:exported="true">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
 
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+        <activity
+            android:name=".MainActivity"
+            android:exported="true">
```
Comment (priyanshudesai) [inline app/src/main/AndroidManifest.xml:29]: if activity is not launcher then why it is exported?

File: app/src/main/java/com/simform/news/data/model/News.kt [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    @SerializedName("status")
+    val status: String,
+    @SerializedName("totalResults")
+    val totalResults: Int
+) : Parcelable
+
+@Parcelize
+data class Article(
+    @SerializedName("author")
+    val author: String? = null,
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/data/model/News.kt:20]: create different file for different model classes

File: app/src/main/java/com/simform/news/data/model/SourceResponse.kt [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    @SerializedName("language")
+    val language: String,
+    @SerializedName("name")
+    val name: String,
+    @SerializedName("url")
+    val url: String
+)
+
+data class SourceResponse(
+    @SerializedName("sources")
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/data/model/SourceResponse.kt:24]: ^^^

File: app/src/main/java/com/simform/news/network/ApiClient.kt [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+import okhttp3.OkHttpClient
+import okhttp3.logging.HttpLoggingInterceptor
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+
+object ApiClient {
+
+    val apiInterface: ApiInterface by lazy {
+        val interceptor = HttpLoggingInterceptor()
+
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/network/ApiClient.kt:13]: where have you set level of interceptor?

File: app/src/main/java/com/simform/news/repository/HeadlineRepository.kt [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+    fun getNews(category: String, callback: (News) -> Unit) {
+        val response = ApiClient.apiInterface.getNews(category, URL.apiKey)
+        response.enqueue(object : Callback<News> {
+            override fun onResponse(call: Call<News>, response: Response<News>) {
+                if (response.isSuccessful) {
+                    response.body()?.let { callback(it) }
+                } else {
+
+                }
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/repository/HeadlineRepository.kt:20]: what will be done here in else block?

File: app/src/main/java/com/simform/news/repository/HeadlineRepository.kt [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+            }
+
+            override fun onFailure(call: Call<News>, t: Throwable) {
+
+            }
+        })
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/repository/HeadlineRepository.kt:26]: ^^^

File: app/src/main/java/com/simform/news/repository/SourcesRepository.kt [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+import com.simform.news.data.model.SourceResponse
+import com.simform.news.network.ApiClient
+import com.simform.news.util.URL
+
+class SourcesRepository {
+
+    suspend fun getSources(): SourceResponse? {
+        val response = ApiClient.apiInterface.getSources("sources", URL.apiKey)
+        val result = response.body()
+        return result
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/repository/SourcesRepository.kt:12]: code can be optimized

File: app/src/main/java/com/simform/news/ui/bookmark/view/BookMarkFragment.kt [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+        binding.rvBookmark.apply {
+            layoutManager = LinearLayoutManager(context)
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/ui/bookmark/view/BookMarkFragment.kt:51]: can we set this layout manger using xml?

File: app/src/main/java/com/simform/news/ui/headlines/viewmodel/DataViewModel.kt [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package com.simform.news.ui.headlines.viewmodel
+
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.simform.news.data.model.Article
+
+class DataViewModel : ViewModel() {
+    val dataList: MutableLiveData<Article> = MutableLiveData()
+}
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/ui/headlines/viewmodel/DataViewModel.kt:9]: remove unused viewmodel

File: app/src/main/java/com/simform/news/ui/home/viewmodel/HomeVM.kt [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package com.simform.news.ui.home.viewmodel
+
+import androidx.lifecycle.ViewModel
+
+class HomeVM : ViewModel() {
+
+}
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/ui/home/viewmodel/HomeVM.kt:7]: ^^^

File: app/src/main/java/com/simform/news/ui/newsdetails/view/ArticleFragment.kt [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+//        }
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/ui/newsdetails/view/ArticleFragment.kt:92]: remove commented code or add specific comments

File: app/src/main/java/com/simform/news/ui/sources/view/SourcesFragment.kt [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    private fun loadNews() {
+        binding.pbLoader.visibility = View.VISIBLE
+
+        viewModel.getSources()
+        viewModel.resources.observe(viewLifecycleOwner) { sourceArrayList ->
+            if (isConnection) {
+                binding.pbLoader.visibility = View.VISIBLE
+            } else {
+                binding.pbLoader.visibility = View.GONE
+            }
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/ui/sources/view/SourcesFragment.kt:82]: can this code be optimised using if else expression

File: app/src/main/java/com/simform/news/util/NetworkStateViewModel.kt [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+                super.onUnavailable()
+//                Toast.makeText(application, "onUnavailable", Toast.LENGTH_SHORT).show()
+                Log.d(TAG, "onUnavailable")
+                networkStateLiveData.postValue(false)
+            }
+
+
+            override fun onLost(network: Network) {
+                super.onLost(network)
+//                Toast.makeText(application, "Lost", Toast.LENGTH_SHORT).show()
```
Comment (priyanshudesai) [inline app/src/main/java/com/simform/news/util/NetworkStateViewModel.kt:48]: ^^^

File: app/src/main/res/layout/fragment_sources.xml [hunk 23]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools">
+
+    <data>
+
+    </data>
+
```
Comment (priyanshudesai) [inline app/src/main/res/layout/fragment_sources.xml:9]: remove unused data tag

**Review by priyanshudesai — CHANGES_REQUESTED** [review]

**Pull request merged**